module/spotinst: Add Spotinst as a new cloud provider

talzur opened the pull request:
##### SUMMARY
This adds a [Spotinst](https://spotinst.com/) module to Ansible, which can manage [Elastigroups](https://help.spotinst.com/hc/en-us/articles/115003530285-Ansible-).

##### ISSUE TYPE
 - New Module Pull Request

### Usage

Create an Elastigroup using the following YAML:

```yaml
- hosts: localhost
  tasks:
    - name: create elastigroup
      spotinst_aws_elastigroup:
          state: present 
          risk: 100
          availability_vs_cost: balanced
          availability_zones:
            - name: us-west-2a
              subnetId: subnet-2b68a15c
          image_id: ami-f173cc91
          key_pair: spotinst-oregon
          max_size: 15
          min_size: 0
          target: 0
          unit: instance
          monitoring: True
          name: ansible-group
          on_demand_instance_type: c3.large
          product: Linux/UNIX
          load_balancers:
            - tal-lb-1
          security_group_ids:
            - sg-8f4b8fe9
          spot_instance_types:
            - c3.large
          state: absent
          do_not_update:
            - image_id
            - target
      register: result
    - debug: var=result
```

talzur committed:
module/spotinst: Add Spotinst as a new cloud provider to manage Elastigroups

talzur committed:
module/spotinst: added necessary documentation and example usage

talzur committed:
module/spotinst: fixed invalid YAML documentation and example usage

talzur committed:
module/spotinst: fixed iam_role documentation and example usage, and added required licensing comment

talzur committed:
module/spotinst: fail aws_elastigroup module if the spotinst SDK is not installed

talzur committed:
module/spotinst: changes to yaml formatting

talzur committed:
module/spotinst: changes to yaml formatting

talzur committed:
module/spotinst: changes to documentation

talzur committed:
module/spotinst: Added the version_added field

talzur committed:
module/spotinst: Added the version_added field

talzur committed:
module/spotinst: Added the version_added field

talzur committed:
module/spotinst: missing documentation added

talzur committed:
module/spotinst: missing documentation field descriptors added

talzur committed:
module/spotinst: fixed documentation line length

talzur committed:
module/spotinst: fixed documentation line length

talzur committed:
module/spotinst: added return values to documentation

talzur committed:
module/spotinst: removed version_added from ANSIBLE_METADATA

talzur committed:
module/spotinst: support health_check_unhealthy_duration_before_replacement

ahharu commented:
@talzur . Please allow the module to get the credentials from environment variables. This will facilitate integration with CI

talzur committed:
module/spotinst: support creds retrieval (spotinst token) from environment variable SPOTINST_TOKEN

talzur commented:
@ahharu Added support for setting the token in environment variables.

talzur committed:
module/spotinst: remove empty lines

ahharu commented:
API expects strategy.fallbackToOd , python spotinst client is sending strategy.fallbackToOndemand ..

= :bomb:

ahharu commented:
To mantain interoperability with ec2_asg I would modify "expand_tags" function as follows

```
def expand_tags(eg_launchspec, tags):
    if tags is not None:
        eg_tags = []

        for tag in tags:
            eg_tag = spotinst.aws_elastigroup.Tag()
            if tag.keys():
                eg_tag.tag_key = tag.keys()[0]
            if tag.values():
                eg_tag.tag_value = tag.values()[0]

            eg_tags.append(eg_tag)

        if eg_tags.__sizeof__() > 0:
            eg_launchspec.tags = eg_tags
```

ahharu commented:
Spotinst python client expects fallback_to_od and the module uses fallback_to_ondemand

please check

talzur committed:
module/spotinst: fixed fallback_to_od, changed tag expansion to conform to ec2_asg

ahharu commented:
@talzur i modified your python script so we can use rolling_update

https://gist.github.com/ahharu/c40b5330a88084aab749c85901736c40

talzur committed:
module/spotinst: fixed tags documentation

talzur commented:
@ahharu 
Fixed fallback_to_od. 
Changed expand_tags method as suggested. Thank you!

As for the rolling update - Great!
Will modify a little and add to the request in a few days, there is an internal convention we want to conform to.

talzur committed:
module/spotinst: fixed tags documentation

talzur committed:
module/spotinst: fixed dictionary field type setters missing

ahharu commented:
@talzur Great! Waiting for it! Also we would like to see any way of configuring the max retries of the elastigroup to launch instances

talzur committed:
module/spotinst: Added roll_config to support rolling the group after update
module/spotinst: Added uniqueness_by option which allows to use non-unique group names. In that case updating and deleting requires setting the id property

talzur committed:
module/spotinst: documentation change

talzur commented:
@ahharu Ok, we now support group roll after an update. 
use roll_config. More info in the documentation.
Will look into the max_retries functionality.

talzur committed:
module/spotinst: added more usage examples and fixed some documentation errors

talzur committed:
module/spotinst: changed metadata

talzur committed:
module/spotinst: changed metadata

talzur committed:
module/spotinst: changed metadata

talzur commented:
ready_for_review

talzur committed:
module/spotinst: return okay when attempting to delete non-existent group in uniqueness_by id mode

talzur commented:
bot_status

talzur committed:
module/spotinst: use module.fail_json instead of raise exception

mikedlr commented on lib/ansible/modules/cloud/spotinst/spotinst_aws_elastigroup.py:
```diff
@@ -0,0 +1,1598 @@
+#!/usr/bin/python
+# This file is part of Ansible
```
There's a new one line copyright statement for modules.  I think you should move to something like that.   See https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#copyright

talzur replied:
Fixed this. Thanks!

mikedlr commented on lib/ansible/modules/cloud/spotinst/spotinst_aws_elastigroup.py:
```diff
@@ -0,0 +1,1598 @@
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'certified'}
```
Should this really be certified already?  Most new modules start as community support then get upgraded?  Not sure what the process is here though.

talzur replied:
Changed to community. Is this okay?

mikedlr commented on lib/ansible/modules/cloud/spotinst/spotinst_aws_elastigroup.py:
```diff
@@ -0,0 +1,1598 @@
+DOCUMENTATION = """
+---
+module: spotinst_aws_elastigroup
+version_added: 2.4
```
2.5

talzur replied:
Fixed.

mikedlr commented on lib/ansible/modules/cloud/spotinst/spotinst_aws_elastigroup.py:
```diff
@@ -0,0 +1,1598 @@
+module: spotinst_aws_elastigroup
+version_added: 2.4
+short_description: Create, update or delete Spotinst AWS Elastigroups
+author: Spotinst
```
Normally your full name here. Not sure what the policy is on commercially supported modules.

mikedlr commented on lib/ansible/modules/cloud/spotinst/spotinst_aws_elastigroup.py:
```diff
@@ -0,0 +1,1598 @@
+        snapshot_id(Integer),
+        volume_type(String),
+        volume_size(Integer))
+    required: false
```
Delete "required: false" in every case;  it is the default so you don't need to state it.

talzur replied:
Fixed.

mikedlr commented on lib/ansible/modules/cloud/spotinst/spotinst_aws_elastigroup.py:
```diff
@@ -0,0 +1,1598 @@
+        availability_vs_cost=dict(type='str'),
+        availability_zones=dict(type='list'),
+        block_device_mappings=dict(type='list'),
+        chef=dict(type='dict', required=False, default=None),
```
`required=False` and `default=None` are defaults so you can eliminate them in every case below

talzur replied:
Fixed.

mikedlr commented on lib/ansible/modules/cloud/spotinst/spotinst_aws_elastigroup.py:
```diff
@@ -0,0 +1,1598 @@
+    module.exit_json(changed=has_changed, group_id=group_id, message=message, instances=instances)
+
+
+def retrieve_creds():
```
I would move this above `def main():` to match with standard practice

talzur replied:
Fixed.

mikedlr commented on lib/ansible/modules/cloud/spotinst/spotinst_aws_elastigroup.py:
```diff
@@ -0,0 +1,1598 @@
+    home = expanduser("~")
+    vars = dict()
+
+    with open(home + "/.spotinst/credentials", "r") as creds:
```
IMHO there should be a possibility to use a different location for the credentials.  This will be important for the integration tests and could be read from the environment variables and/or from the configuration in the playbook.

talzur replied:
I agree, working on this change and all other changes mentioned. Will update soon.

ryansb replied:
Also you'll want to use a `path` type on that parameter, so that Ansible's pre-validation on paths can be used.

talzur replied:
Fixed. Now supports adding a token and account id through env variables, credentials file, and credentials file with different path. The parameter has the path type configured for ansible's pre-validation as well.

mikedlr commented on lib/ansible/modules/cloud/spotinst/spotinst_aws_elastigroup.py:
```diff
@@ -0,0 +1,1598 @@
+        eg.scaling = eg_scaling
+
+
+def expand_scaling_policies(scaling_policies):
```
Generally I'm not so happy with this long list of if statements.  What I would rather have is a set of named objects (like the `eg.policy`) and a set of lists of relevant parameters (`eg_policy_params=['policy_name', 'namespace....`) and then iterate through the list to build the object from the original policy object.  I think this would be easier to maintain in future.

talzur replied:
Great idea. I have implemented this as you suggested. In some cases the spotinst and module name are not identical to conform to aws usage which users are familiar with, as is the case with min_size (As opposed to 'minimum' in spotinst). For these cases I have also allowed to provide a dictionary in the list with the values for the spotinst and ansible names, and have taken it into account in the expand_fields method which populates the objects.

mikedlr reviewed (changes requested):
Various comments; I'll discuss a bit more on IRC.

mikedlr commented on lib/ansible/modules/cloud/spotinst/spotinst_aws_elastigroup.py:
```diff
@@ -0,0 +1,1598 @@
+    wait_for_instances = module.params.get('wait_for_instances')
+
+    if wait_timeout is None:
+        wait_timeout = 300
```
In the integration tests related to RDS I have had to increase timeouts for creation to up to 800.  I had some similar problems working on the auto scaling groups earlier.  I would propose that this timeout should be at the level where the wait is sufficent 95% of the time even at times when AWS is reasonably busy and instance start up takes lots of time.  Do you know if 300 is long enough?

ryansb replied:
Yeah, for spot instances I'd say 300 is a bit short, unless spotinst behaves *significantly* differently from the AWS spot instance market.

mikedlr commented:
@talzur you missed my IRC comment (you might want to get an IRC bouncer or use one of the services that lets you get it to your mobile).  Here is a cut and paste.  

```
⇐ TalZ quit  
11:09:43 M<mikedlr> I have some example of something similar in the rds instance split module - https://github.com/ansible/ansible/pull/26602  - there must be better examples
11:12:07 M<mikedlr> the validate_parameters(required_vars, valid_vars, module) function picks out the parameters needed at each API call.
11:37:27 M<mikedlr> I'm rewriting it as select_parameters(module, required_vars, valid_vars) in the new version but that's a bit broken https://github.com/ansible/ansible/pull/29032/files
11:37:46 M<mikedlr> oops..  he quit again ;-)
```

sumeshpremraj commented on lib/ansible/modules/cloud/spotinst/spotinst_aws_elastigroup.py:
```diff
@@ -0,0 +1,1598 @@
+
+  tags:
+    description:
+      - (Dictionary) a dictionary of tags to configure in the elastigroup. Please specify list of keys and values (key colon value);
```
Tags is compatible with ec2_asg module and now takes a **list**, so this should be updated.

sumeshpremraj commented:
@talzur can you help me understand what the following attributes in up and down_scaling_policies are:
- `operator` (String, required)
      I guess this is `>` or `<`?
- `action_type` (String, required)

**Update (for anyone else looking):**
Here's a reply from spotinst support that helped me fix the issue:
> While using action_type, please also set the following: 
> In case of adjustment - set the field "adjustment"
> In case of setMaxTarget - set the field "max_target_capacity"
> In case of setMinTarget - set the field "min_target_capacity"

```         up_scaling_policies:
            - ...
              operator: gt
              action_type: adjustment
              adjustment: 2
```


Hope this is updated in Spotinst website and inside the module soon.

talzur committed:
module/spotinst: changed wrong declaration of state variable in documentation example, and fixed policy action usage

ahharu commented:
wait_for_instances should also wait for a rolling update to finish..

ahharu commented:
I solved with this but it would be good to update the "spotinst" pypi module to make this cleaner 

`    roll_url = "https://api.spotinst.io/aws/ec2/group/{}/roll?limit=1".format(group_id)

    if state == 'present' and group_id is not None and rolling_update is not None:
        # Rolling update. Wait for it, it's going to be legendary!
        wait_timeout = time.time() + ( wait_timeout * target_instances ) 

        result = client.send_get(roll_url,"elastigroup_roll")
        
        if len(result['response']['items']) == 0:
            return []
        
        status = 'waiting'
        while wait_timeout > time.time() and status != 'finished':
            last_roll_id = result['response']['items'][0]['id']
            roll_status_url = "https://api.spotinst.io/aws/ec2/group/{}/roll/{}".format(group_id,last_roll_id)
            
            status = client.send_get(roll_url,"elastigroup_roll")['response']['items'][0]['status']
            
            time.sleep(10)
`

talzur committed:
module/spotinst: fix scale_target_capacity, scale_min_capacity, and scale_max_capacity being ignored

talzur committed:
module/spotinst: Added examples for handling network interfaces

sumeshpremraj commented:
@talzur thanks for all your work on this, it made using spotinst easier than importing everything manually.

Small request - can you make spotinst module a drop in replacement for ec2_asg module (in terms of attribute names and expected values)? Currently, there are a lot of changes which forced me to duplicate a lot of attributes and tweak them slightly, this makes it harder for someone to start using elastigroups quickly, or use this along with ec2_asg (as a fallback), like I do.

A couple of examples off the top of my head: `block_device_mappings`, the scaling policy `operator`.

talzur committed:
module/spotinst/spotinst_aws_elastigroup: one line copyright statement for modules
module/spotinst/spotinst_aws_elastigroup: changed supported_by value to community
module/spotinst/spotinst_aws_elastigroup: changed version added to 2.5
module/spotinst/spotinst_aws_elastigroup: removed required:false as it is the default
module/spotinst/spotinst_aws_elastigroup: iterate through object fields to create an object instead of having an if statement per parameter
module/spotinst/spotinst_aws_elastigroup: eliminated required=False and default=None in all module fields properties. Added required=True where necessary.
module/spotinst/spotinst_aws_elastigroup: added the path property to the credentials_path variable, to allow for ansible's pre-validation on paths
module/spotinst/spotinst_aws_elastigroup: added the possibility to use a different location for the credentials
module/spotinst/spotinst_aws_elastigroup: updated the tags documentation
module/spotinst/spotinst_aws_elastigroup: Added target tracking policy support
module/spotinst/spotinst_aws_elastigroup: added target tracking example usage

talzur committed:
module/spotinst/spotinst_aws_elastigroup: formatting

pilou- commented on lib/ansible/modules/cloud/spotinst/spotinst_aws_elastigroup.py:
```diff
@@ -0,0 +1,1511 @@
+    The credentials file must contain a row that looks like this
+    token = <YOUR TOKEN>
+    Full documentation available at https://help.spotinst.com/hc/en-us/articles/115003530285-Ansible-
+options:
```
documentation for `credentials_path` is missing.

talzur replied:
Right. Fixed. Thanks!

talzur committed:
module/spotinst/spotinst_aws_elastigroup: formatting
module/spotinst/spotinst_aws_elastigroup: deprecated usage of os.path expanduser

talzur committed:
spotinst/ansible-module: Added credentials_path documentation

talzur committed:
spotinst/ansible-module: Removed unused import

ryansb commented on lib/ansible/modules/cloud/spotinst/spotinst_aws_elastigroup.py:
```diff
@@ -0,0 +1,1509 @@
+  - Can create, update, or delete Spotinst AWS Elastigroups
+    Launch configuration is part of the elastigroup configuration,
+    so no additional modules are necessary for handling the launch configuration.
+    This module requires that you install the spotinst Python SDK (pip install spotinst)
```
There is a `requirements:` docs key where you can list these as well, and optionally include a specific version https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/aws_s3_bucket_facts.py#L17

talzur replied:
Thanks, changed this.

ryansb commented on lib/ansible/modules/cloud/spotinst/spotinst_aws_elastigroup.py:
```diff
@@ -0,0 +1,1509 @@
+    This module requires that you install the spotinst Python SDK (pip install spotinst)
+    You will have to have a credentials file in this location - <home>/.spotinst/credentials
+    The credentials file must contain a row that looks like this
+    token = <YOUR TOKEN>
```
I think this should also be available as a direct option. Many Ansible users make use of `ansible-vault` where they can store secrets as variables instead of plaintext in a config. For example, a user might prefer to use Vault and:

```
- spotinst_aws_elastigroup:
    token: "{{ sp_inst_token }}"
    availability_zones: [......]
    .....
```

talzur replied:
I agree. Great idea. I have also changed the token and account specificity as you've suggested in another comment.

ryansb commented on lib/ansible/modules/cloud/spotinst/spotinst_aws_elastigroup.py:
```diff
@@ -0,0 +1,1509 @@
+    You will have to have a credentials file in this location - <home>/.spotinst/credentials
+    The credentials file must contain a row that looks like this
+    token = <YOUR TOKEN>
+    Full documentation available at https://help.spotinst.com/hc/en-us/articles/115003530285-Ansible-
```
Awesome, thanks for making such a comprehensive post 👍

ryansb commented on lib/ansible/modules/cloud/spotinst/spotinst_aws_elastigroup.py:
```diff
@@ -0,0 +1,1509 @@
+except ImportError:
+    pass
+
+eni_fields = ['description',
```
Since all of these appear to be constants, it seems to me that making them tuples instead of lists would be good to avoid accidentally modifying them.

talzur replied:
Done 👍

ryansb commented on lib/ansible/modules/cloud/spotinst/spotinst_aws_elastigroup.py:
```diff
@@ -0,0 +1,1509 @@
+
+    account = os.environ.get('ACCOUNT')
+    if not account:
+        account = creds_file_loaded_vars.get("account")
```
This would put the account in the creds file *above* the params directly passed to the module. This should probably be reversed, as we try to follow the order of specificity (preferring most specific) as: /etc file, $HOME file, environment variable, direct specification on task parameter.

ryansb commented on lib/ansible/modules/cloud/spotinst/spotinst_aws_elastigroup.py:
```diff
@@ -0,0 +1,1509 @@
+        if group['name'] == name:
+            group_found = True
+            group_id = group.get('id')
+            break
```
I think these 3 lines can be replaced with `return True, group.get('id')` and the below return can change to `return False, ''`

talzur replied:
Very elegant! Made this change as well.

ryansb commented on lib/ansible/modules/cloud/spotinst/spotinst_aws_elastigroup.py:
```diff
@@ -0,0 +1,1509 @@
+            group_id = external_group_id
+    else:
+        groups = client.get_elastigroups()
+        group_found, group_id = find_group_with_same_name(groups, name)
```
Can `group_found` just be swapped to `should_create` to get rid of the below `if`?

ryansb commented on lib/ansible/modules/cloud/spotinst/spotinst_aws_elastigroup.py:
```diff
@@ -0,0 +1,1509 @@
+        if state == 'present':
+            eg = expand_elastigroup(module, is_update=False)
+
+            processLog = open("/var/log/spotinst/ansible-module.log", 'a+')
```
Instead of manually creating log files such as this, it's preferred to use the `module.debug(.........)` method so that the Ansible controller is in charge of deciding whether the logs should be displayed.

talzur replied:
👍

ryansb commented on lib/ansible/modules/cloud/spotinst/spotinst_aws_elastigroup.py:
```diff
@@ -0,0 +1,1509 @@
+
+            eg_bdms.append(eg_bdm)
+
+        if eg_bdms.__sizeof__() > 0:
```
Please use `if len(eg_bdms) > 0` instead of directly calling `__sizeof__`

talzur replied:
Changed all sizeof to len()

ryansb commented on lib/ansible/modules/cloud/spotinst/spotinst_aws_elastigroup.py:
```diff
@@ -0,0 +1,1509 @@
+
+            eg_enis.append(eg_eni)
+
+        if eg_enis.__sizeof__() > 0:
```
Same - please use `len`

ryansb commented on lib/ansible/modules/cloud/spotinst/spotinst_aws_elastigroup.py:
```diff
@@ -0,0 +1,1509 @@
+
+    if up_scaling_policies is not None:
+        eg_up_scaling_policies = expand_scaling_policies(up_scaling_policies)
+        if eg_up_scaling_policies.__sizeof__() > 0:
```
Same - please use `len`

ryansb commented on lib/ansible/modules/cloud/spotinst/spotinst_aws_elastigroup.py:
```diff
@@ -0,0 +1,1509 @@
+
+    if down_scaling_policies is not None:
+        eg_down_scaling_policies = expand_scaling_policies(down_scaling_policies)
+        if eg_down_scaling_policies.__sizeof__() > 0:
```
Same - please use `len`

ryansb commented on lib/ansible/modules/cloud/spotinst/spotinst_aws_elastigroup.py:
```diff
@@ -0,0 +1,1509 @@
+
+    if target_tracking_policies is not None:
+        eg_target_tracking_policies = expand_target_tracking_policies(target_tracking_policies)
+        if eg_target_tracking_policies.__sizeof__() > 0:
```
Same - please use `len`

talzur committed:
spotinst/ansible-module: Added requirements field
spotinst/ansible-module: Changed field lists to immutable tuples
spotinst/ansible-module: simplified existing group finding method
spotinst/ansible-module: using module.debug instead of process log
spotinst/ansible-module: changed usage of _sizeof_ to len() where applicable
spotinst/ansible-module: token can be directly set as parameter
spotinst/ansible-module: account and token set through module params are preferred, then env vars, and lastly through file, to conform to the order of specificity

ryansb commented on lib/ansible/modules/cloud/spotinst/spotinst_aws_elastigroup.py:
```diff
@@ -0,0 +1,1507 @@
+             'subnet_id',
+             'placement_group_name')
+
+opsworks_fields = ('layer_id')
```
For these to be a tuple you need a trailing comma, like `opsworks_fields = ('layer_id',)` for all single-item tuples.

talzur replied:
Thanks, fixed.

talzur committed:
spotinst/ansible-module: removed iam role validation
spotinst/ansible-module: changed all single item field tuples to proper tuples

talzur committed:
spotinst/ansible-module: fixed action_type & type discrepancy
spotinst/ansible-module: empty instances list on start of retrieval to avoid duplication of instances
spotinst/ansible-module: fix IOError when no creds file is set

talzur committed:
spotinst/ansible-module: formatting

talzur committed:
spotinst/ansible-module: only expand role if one of the fields is set

s-hertel commented:
Thanks @talzur!

talzur commented:
Thanks @s-hertel & @ryansb !
